apiserver/facades/agent/secretsdrain/register.go

```diff
@@ -23,27 +23,32 @@ func Register(registry facade.FacadeRegistry) {
 }
 
 // newSecretsDrainAPI creates a SecretsDrainAPI.
-func newSecretsDrainAPI(context facade.ModelContext) (*commonsecrets.SecretsDrainAPI, error) {
-	if !context.Auth().AuthUnitAgent() && !context.Auth().AuthApplicationAgent() {
+func newSecretsDrainAPI(ctx facade.ModelContext) (*commonsecrets.SecretsDrainAPI, error) {
+	if !ctx.Auth().AuthUnitAgent() {
 		return nil, apiservererrors.ErrPerm
 	}
-	leadershipChecker, err := context.LeadershipChecker()
+	leadershipChecker, err := ctx.LeadershipChecker()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	model, err := context.State().Model()
+	model, err := ctx.State().Model()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	authTag := context.Auth().GetAuthTag()
+	authTag := ctx.Auth().GetAuthTag()
+
+	cloudService := ctx.ServiceFactory().Cloud()
+	credentialSerivce := ctx.ServiceFactory().Credential()
+	secretBackendAdminConfigGetter := func(stdCtx context.Context) (*provider.ModelBackendConfigInfo, error) {
+		return commonsecrets.AdminBackendConfigInfo(stdCtx, commonsecrets.SecretsModel(model), cloudService, credentialSerivce)
+	}
 	return commonsecrets.NewSecretsDrainAPI(
 		authTag,
-		context.Auth(),
-		context.Logger().Child("secretsdrain"),
+		ctx.Auth(),
+		ctx.Logger().Child("secretsdrain"),
 		leadershipChecker,
 		commonsecrets.SecretsModel(model),
-		state.NewSecrets(context.State()),
-		context.State(),
-		context.WatcherRegistry(),
+		ctx.ServiceFactory().Secret(secretBackendAdminConfigGetter),
```

Comment:
Doesn't seem correct to put the backend config getter globally.
The user secret and charm secret are in the same `secret` service?

Reply:
This is used until the secret backend service is fully implemented. It is needed for when the secret is stored in a backend (not state) and the secret is deleted - the secret service will get the backend to also delete the content. There's a bunch of logic for dealing with backend content still in the facades which all needs to be reworked once the backend service is in place.